### ISSUE TITLE
First byte used to represent the encoding or settle on base58bitcoin-multihash?

### ISSUE THREAD

**Issue by benjaminjackman** [multiformats/multihash#14]
There probably aren't going to be that many different encodings, however consider if the first byte (shown as an ASCII char in '') represented the encoding:
e.g. 
```
'1' or 0x31 = Binary #Just put this here for fun really
'2' or 0x32 = Hex
'3' or 0x33 = Base32
'4' or 0x34 = Base36 #Case insensitive letters and numbers
'5' or 0x35 = Base58Bitcoin #Guessing bitcoin is the alphabet (e.g. not Ripple)?
'6' or 0x36 = Base64
```
etc.

So the examples from the page would change to be
```
# sha1 - 0x11
211140beec7b5ea3f0fdbc95d0dd47f3c5bc275da8a33 # sha1 in hex
324a0qvp7pqn3y3yvt5egvn3z7hdw4xeuh8tg # sha1 in base32
55dqx43zNtUUbPj97vJhpHyUUPyrmXG # sha1 in base58
6ERQL7se16j8P28ldDdR/PFvCddqKMw== # sha1 in base64

# sha2-256 0x12
212202c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae # sha2-256 in hex
328g2r9nmddmfzhmfz6dmaf0x610k84u25nr690xzm3wrmqm8c9kefbg # sha256 in base32
5QmRJzsvyCQyizr73Gmms8ZRtvNxmgqumxc2KUp71dfEmoj # sha256 in base58
6EiAsJrRraP/Gj/mbRTwdMEE0E0ItcGSDv6D5il6IYmbnrg== # sha256 in base64
```

Then the multihash would be truly self-describing, instead of settling on what appears to be a de-facto base58 encoding. It would also eliminate the need for the -e parameter to the hashpipe program.

I didn't work through it, but looking over all the currently generated multihashes in existence, it might be possible to implement this in a backwards compatible way, if clever characters are chosen for the different encodings such that they don't intersect with possible numbers in the wild (e.g. 1 is out because hex has it for SHAs, Q is out etc).

That of multihash might just mean `base58` in any case either the encoding should be put into the multihash or different encodings probably ought to be dropped altogether to avoid confusion and have a universal scheme, maybe some superset of the multihash called the unihash (then the multihash could be used for hash name that really do contain multiple hashes, something this doesn't quite support, but which could easily).

**Comment by jbenet** [comment]
So, i really like this idea in general. thank you for posting here. I really wanted to do something like this but couldn't figure out how to make it work, so i shipped the standard without it. the constraints i wanted to fill were:

- have a way of having the encoding character by fine in that encoding. meaning, say, that the bits used to define hex, would be printed in valid hex. this can be a hard thing to do, i haven't found but i haven't explored much for failure cases.
- have a way to agree on alphabets. not all base{32, 58, 64}s are equal.
- have a way to extend this to lots of other encodings. (though admittedly, maybe just focusing on the number bases would probably be best. and [multicodec](https://github.com/jbenet/multicodec) has emerged to handle serialization encodings, so it is probably fine to just do bases here)

> Then the multihash would be truly self-describing, instead of settling on what appears to be a de-facto base58 encoding. It would also eliminate the need for the -e parameter to the hashpipe program.

yes this would be a huge win,

> I didn't work through it, but looking over all the currently generated multihashes in existence, it might be possible to implement this in a backwards compatible way, if clever characters are chosen for the different encodings such that they don't intersect with possible numbers in the wild (e.g. 1 is out because hex has it for SHAs, Q is out etc).

yeah, i would really want this change to be backwards compatible. it doesn't _have_ to be in the standard, but IPFS nodes would definitely have to handle multihashes without the encoding byte, to ensure no links break. your suggestion (i think) is incompatible with my requirement above (that the encoding be printed in the encoding itself, too).

**Comment by benjaminjackman** [comment]
So which encodings are currently being used in actual multihashes right now?

The more I think about it, the more I think it should just be set to base58bitcoin and, maybe, HEX as well. 
base58bitcoin has almost the same size compression benefits of something like base64.

I say maybe on HEX because it's:
+ Easy to read because the bytes all line up nicely.
+ Easy to write code for.
+ HEX+SHA1 = git+an easily removable 2-3 byte prefix
- Takes up a lot more space.

I think all the other encodings should be dropped and left as future engineering challenges (to avoid overengineering now).

Overall 
+ it considerably easier to write a multihash encoder / decoder.
+ moves towards a clear 'one way' pythonesque philosophy way to do things
+ removes ambiguity as to what the encoding is for a given a multihash
- multihash doesn't support quite as many encodings (but does that matter?).

What really matters is having the multihash be resilient to *digests* falling in and out of favor.

If the encoding space is limited to HEX and base58bitcoin, and practically all current multihashes in current existence are base58bitcoin we could define a HEX / base58 bit on the byte currently used for encoding functions and have that bit toggle between Hex and base58, or just say that hex starts with a character that base58 can't currently emit as the first byte given the digests currently in use.

**Comment by jbenet** [comment]
> The more I think about it, the more I think it should just be set to base58bitcoin and, maybe, HEX as well.  base58bitcoin has almost the same size compression benefits of something like base64.

base58 is _such_ a pain because it has all this expensive math. we're only defaulting to it because base64 cannot be made copy-pastable easily. (need 2 more alphanums in ascii). have thought about using other characters that are alphanum in utf-8 (like o's with accents, or something), but this makes them hard to type...

i would really like to find a good way to support the standard bases (16, 32, 64) too. it's just somewhat tricky. 

((oh i also thought about a base1024 and base2048 using utf8 chars, there's benefit to super screen-compressed copy-pastable strings, even if these are really inefficient in true bit size)

**Comment by jbenet** [comment]
maybe:
```
base1         1    unary tends to be 11111 
base2         0    binary has 1 and 0
base8         7    highest char in octal
base16        Ff   highest char in hex
base32        Uu   rfc4648 - highest letter
base32hex     Vv   rfc4648 - highest char
base58flickr  Z    highest char
base58btc     z    highest char
base64        y    rfc4648 highest char
base64url     Y    rfc4648 highest char
base128      
base256ascii  X    ascii
base-emoji    😎   base emoji
base65536     ᔰ    base65536 0
```

**Comment by jbenet** [comment]
Thoughts on the above o/ ?

**Comment by Kubuxu** [comment]
There is implementation of base32 that can be used in DNS https://dnscurve.org/in-implement.html, it is used in cjdns and it would be nice to support it also, it is not critical though.

**Comment by mildred** [comment]
I don't like much this idea because I don't think that's in scope for the multihash. I really like the multihash as it is and I fear that by adding the encoding we are adding too much things.

I had the idea that a multihash was always a binary string, like any hash digest we get out there. And as with any digest, there are different ways to encode it when we transmit it. These different encodings are specific to the way we transmit them:

- if we have a binary protocol, we want to include it in binary and we don't want to have to parse it from base64 or any other textual format

- if we pass around a path / url, we want a textual format (the format depending on how the URL or path prefix has been specified). We absolutely don't want a binary string.

It also gives multiple text representations for the same path, in case we have a hash as a path component. We must decode the path to normalize it to a standard encoding if we want to compare paths.

Also, in the case we wanted to encode the path in binary (for example in ipfs/specs#61) do we have a way to detect the hash length so we know when to stop parsing the hash?

However, I like the idea from a user interface perspective, but the protocols and lower layers should not have to support this. Base58 is already nice because there is only one way to encode (there is no lowercase/uppercase problem of hexadecimal). So perhaps, instead of making this part of multihash, it should be in a separate layer that would be used by the user facing code only. Not the lower levels. This could be something like the multicodecs, except it would have a 1 byte header instead.

**Comment by jbenet** [comment]
> I had the idea that a multihash was always a binary string, like any hash digest we get out there. And as with any digest, there are different ways to encode it when we transmit it.

very much agreed. and i hear you re not adding more to multihash. i just want to make sure we use them correctly:

- always use binary rep when space matters disk + wire
- use base58 when humans + UI matter (copy-paste)
- use base32 or base16 for dns and filesystems, where char limits, cases, or speed matter

Maybe this should really be a "companion" spec. it's like a multicodec but small and specific for bases. (dare i say multibase? or multicodec-base). we could make a section in the multihash spec that describes this use case, linking to whatever defines this thing.

> So perhaps, instead of making this part of multihash, it should be in a separate layer that would be used by the user facing code only. Not the lower levels. This could be something like the multicodecs, except it would have a 1 byte header instead.

Oh perfect!! i wrote the above as i read your post. I think you're totally right and this is likely the best way to go. Do think we should encourage people to use (store, transfer, display) multihash along with this thing (multibase/multicodec-base)

**Comment by eminence** [comment]
The idea of a fully self-describing encoding is attractive, but @mildred analogy to other hash functions (where their outputs is always a fixed-length binary string) is compelling.  

I suspect my desire for a self-describing format is partially due to an observation that we often use the term "multihash" without specifying what encoding is required.  Some examples:

* Within [merkledag.proto](https://github.com/ipfs/go-ipfs/blob/master/merkledag/pb/merkledag.proto) a `PBLink` has a field that is just documented as "multihash of the target object", but I believe it in fact must be the binary encoding.
* Within the API documentation, many arguments are describe as a multihash, but don't specify it requires base58 encoding
* The files in `$IPFS_PATH/blocks` are all hex-encoded multihashes (which might be the only place we hex-encode?)

If you find this to be a weak argument, I probably would agree.  In truth, it's not very hard to figure out exactly what "kind" of multihash to use in each of these contexts.

So in summary, I guess I am +1 to the idea of formally defining a multihash as a function that just produces a binary string, and +1 to the idea of making it clearer what encoding is used (either by making the encodings self-describing or clear via documentation)

**Comment by candeira** [comment]
+1 to multihash values being canonically binary, and other encodings-agnostic.

The encoding can always be part of other multihash-using protocols/standard. So, if a IPFS url is always going to include a bitcoin-alphabet-b58 of a multihash, that would be specified in the IPFS url spec, not in the multihash spec.

I was on the fence about this, in fact the idea of encoding also being part of the hash description was somewhat attractive to me. Thanks to @mildred for helping me make my mind with her arguments. I have only one quibble:

> Base58 is already nice because there is only one way to encode (there is no lowercase/uppercase problem of hexadecimal).

But there are at least three different alphabets for Base58: bitcoin, flickr and ripple.

This, plus @eminence's comment that the IPFS documentation often uses "multihash" without specifying an encoding gives good reasons to decide that:

a) the multihash protocol is encoding-agnostic, and 
b) for the rest of IPFS to standardise on bitcoin's alphabet when using b58, stating it explicitly every time Base58 is mentioned in the documentation.
c) the IPFS documentation should specify an encoding, whether b58, hex, binary or whatever, everytime multihash is part of any specification.

**Comment by jbenet** [comment]
(Will respond more later)

I want to remind that we _already_ have to face the use of multiple
encodings: binary, base58, hex. IPFS stores files on disk (on the repo)
using a hex encoded multihash. This is because (a) Some file systems are
case sensitive!! (Cannot use base58), and (b) hex is way faster to encode
than base58 (or base32) (with base58 at least, encoding hundreds or
thousands of hashes slowed down things. There's places where we opted to
use a raw binary hash instead of a string because it is a tight loop and
the encoding of so many hashes was slowing down ipfs dramatically.

It would be much easier to reason about things if the encoding was always
specified and not left up to all the specs to satisfy.

I do agree it shouldn't necessarily be part of the multihash spec, but I do
think our uses in ipfs should prefer the right encoding for each job, and
to do so without causing major confusion (prefixed).

**Comment by jbenet** [comment]
i still think like this: https://github.com/jbenet/multihash/issues/14#issuecomment-174081767

**Comment by jbenet** [comment]
This is becoming its own multiformat: multibase.
